flamingo/grammar/index.h

```diff
@@ -59,11 +59,18 @@ static int parse_index(flamingo_t* flamingo, TSNode node, flamingo_val_t** val,
 
 	// Prepare result value and exit here if we discard it.
 
-	if (val == NULL) {
+	if (val == NULL && slot == NULL) {
+		val_decref(indexed_val);
+		val_decref(index_val);
 		return 0;
```

Comment:
same comment as above, but this can just replace the `return` after the `assert(false)` at the end of this function, and we can goto there.

Maybe even a pattern like having an `rv = 0` at the top of the function which the last `return` returns, and setting it to `error()` so we can goto the same `val_decref`'s.